drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,514 @@
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <soc.h>
+#define LOG_LEVEL CONFIG_KSCAN_LOG_LEVEL
+LOG_MODULE_REGISTER(kscan_npcx);
+
+#define KEYBOARD_COLUMN_DRIVE_ALL  -2
+#define KEYBOARD_COLUMN_DRIVE_NONE -1
+/* Number of tracked scan times */
+#define SCAN_OCURRENCES		   30U
+
+#define NPCX_KB_ROW_MASK (BIT(CONFIG_KSCAN_NPCX_ROW_SIZE) - 1)
+
+/* Driver config */
+struct kscan_npcx_config {
+	/* keyboard scan controller base address */
+	struct kbs_reg *base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const struct pinctrl_dev_config *pcfg;
+	/* Keyboard scan input (KSI) wake-up irq */
+	int irq;
+	/* Size of keyboard inputs-wui mapping array */
+	int wui_size;
+	/* Mapping table between keyboard inputs and wui */
+	struct npcx_wui wui_maps[];
+};
+
+struct kscan_npcx_data {
+	/* variables in usec units */
+	uint32_t deb_time_press;
+	uint32_t deb_time_rel;
+	int64_t poll_timeout;
+	uint32_t poll_period;
+	uint8_t matrix_stable_state[CONFIG_KSCAN_NPCX_COLUMN_SIZE];
+	uint8_t matrix_unstable_state[CONFIG_KSCAN_NPCX_COLUMN_SIZE];
+	uint8_t matrix_previous_state[CONFIG_KSCAN_NPCX_COLUMN_SIZE];
+	/* Index in to the scan_clock_cycle to indicate start of debouncing */
+	uint8_t scan_cycle_idx[CONFIG_KSCAN_NPCX_COLUMN_SIZE][CONFIG_KSCAN_NPCX_ROW_SIZE];
+	/* Track previous "elapsed clock cycles" per matrix scan. This
+	 * is used to calculate the debouncing time for every key
+	 */
+	uint8_t scan_clk_cycle[SCAN_OCURRENCES];
+	struct k_sem poll_lock;
+	uint8_t scan_cycles_idx;
+	kscan_callback_t callback;
+	struct k_thread thread;
+	atomic_t enable_scan;
+
+	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_KSCAN_NPCX_THREAD_STACK_SIZE);
+};
+
+static struct miwu_dev_callback ksi_callback[CONFIG_KSCAN_NPCX_ROW_SIZE];
+
+/* Keyboard Scan local functions */
+static void kscan_npcx_ksi_isr(const struct device *dev, struct npcx_wui *wui)
+{
+	ARG_UNUSED(wui);
+	struct kscan_npcx_data *const data = dev->data;
+
+	k_sem_give(&data->poll_lock);
+}
+
+static int kscan_npcx_resume_detection(const struct device *dev, bool resume)
+{
+	const struct kscan_npcx_config *const config = dev->config;
+
+	if (resume) {
+		irq_enable(config->irq);
+	} else {
+		irq_disable(config->irq);
+	}
+
+	return 0;
+}
+
+static int kscan_npcx_drive_column(const struct device *dev, int col)
+{
+	const struct kscan_npcx_config *config = dev->config;
+	struct kbs_reg *const inst = config->base;
+	uint32_t mask;
+
+	if (col >= CONFIG_KSCAN_NPCX_COLUMN_SIZE) {
+		return -EINVAL;
+	}
+
+	if (col == KEYBOARD_COLUMN_DRIVE_NONE) {
+		/* Drive all lines to high. ie. Key detection is disabled. */
+		mask = ~0;
+	} else if (col == KEYBOARD_COLUMN_DRIVE_ALL) {
+		/* Drive all lines to low for detection any key press */
+		mask = ~(BIT(CONFIG_KSCAN_NPCX_COLUMN_SIZE) - 1);
+	} else {
+		/*
+		 * Drive one line to low for determining which key's state
+		 * changed.
+		 */
+		mask = ~BIT(col);
+	}
+
+	LOG_DBG("%x", mask);
```

Comment:
maybe add something more to the debug print?

Reply:
Done.